File: 2021/07/17/main.go
package main

func threeEqualParts(arr []int) []int {
	n, s := len(arr), 0

	for i := 0; i < n; i++ {
		s += arr[i]
	}

	if s == 0 {
		return []int{0, n - 1}
	}

	if s%3 != 0 {
		return []int{-1, -1}
	}

	k, j, t, i := s/3, n-1, 0, 0

	for j >= 0 {
		if arr[j] == 1 {
			t += 1
		}
		if t == k {
			break
		}
		j -= 1
	}

	for i < n {
		if arr[i] == 1 {
			break
		}
		i += 1
	}

	for l := j; l < n; l++ {
		if arr[i] != arr[l] {
			return []int{-1, -1}
		}
		i += 1
	}

	k = i
	for k < n {
		if arr[k] == 1 {
			break
		}
		k += 1
	}

	for l := j; l < n; l++ {
		if arr[k] != arr[l] {
			return []int{-1, -1}
		}
		k += 1
	}

	return []int{i - 1, k}
}

File: 2021/07/17/main.py
class Solution:
    def threeEqualParts(self, arr: List[int]) -> List[int]:
        n, s = len(arr), sum(arr)

        if s == 0:
            return [0, n - 1]

        if s % 3:
            return [-1, -1]

        k, j, t, i = s // 3, n - 1, 0, 0
        while j >= 0:
            t += arr[j]==1
            if t == k:
                break
            j -= 1

        while i < n:
            if arr[i]: break
            i += 1

        for l in range(j, n):
            if arr[i] != arr[l]:
                return [-1, -1]
            i += 1
        
        for k in range(i, n):
            if arr[k]: break
        
        for l in range(j, n):
            if arr[k] != arr[l]:
                return [-1, -1]
            k += 1
            
        return [i - 1, k]